### PR TITLE
ZS2022-44 Refactor TagPage Tests

### PR DESCRIPTION
**Pull request by DariusChis**
https://zenitech.atlassian.net/browse/ZS2022-44

### CODE REVIEW

File: src/main/java/driver/DriverFactory.java [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package driver;
 
+import org.openqa.selenium.By;
```
Comment (CaprarClaudiu) [inline src/main/java/driver/DriverFactory.java:3]: :)

File: src/main/java/driver/DriverFactory.java [hunk 2]
```diff
@@ -30,7 +31,7 @@ public DriverFactory(WebDriver driver) {
     @FindBy(id = "filter_device_button")
     private WebElement filterDeviceButton;
     @FindBy(id = "device-filter-apply-button")
-    private WebElement deviceFilterApplyButton;
+    WebElement deviceFilterApplyButton;
```
Comment (CaprarClaudiu) [inline src/main/java/driver/DriverFactory.java]: do we need to change this to be public?

File: src/main/java/pageObjects/TagsPage.java [hunk 3]
```diff
@@ -20,23 +20,21 @@ public class TagsPage extends DriverFactory {
...
+        devicePage = new DevicePage(driver);
+        filter = new Filter(driver);
+        permissions = new Permissions(driver);
     }
 
-    @FindBy(id = "undefined-tags-mButton")
-    private WebElement tagsButtonPage;
+    private DevicePage devicePage;
+    private Filter filter;
+    private Permissions permissions;
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/TagsPage.java]: we should not do this, we should just declare the web elements here also.

File: src/main/java/pageObjects/TagsPage.java [hunk 4]
```diff
@@ -46,160 +44,121 @@ public TagsPage(WebDriver driver) {
...
+        permissions.tagsButtonSidebarMenu.click();
+        elementToLoad(addTagField);
         addTagField.sendKeys(name);
         addTagButton.click();
-        wait.until(ExpectedConditions.visibilityOf(successToaster));
+        elementToLoad(devicePage.successToaster);
         wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("tags-list-tag-" + name + "-item-delete")));
         driver.findElement(By.id("tags-list-tag-" + name + "-item-delete")).click();
-        wait.until(ExpectedConditions.visibilityOf(successToaster));
+        elementToLoad(devicePage.successToaster);
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/TagsPage.java]: we should declare the elements in this class. its better than calling them from other classes especially if they are not from a helper class

File: src/main/java/pageObjects/TagsPage.java [hunk 5]
```diff
@@ -46,160 +44,121 @@ public TagsPage(WebDriver driver) {
...
-        tagsButtonPage.click();
-        wait.until(ExpectedConditions.visibilityOf(addTagField));
+        permissions.tagsButtonSidebarMenu.click();
+        elementToLoad(addTagField);
         addTagField.sendKeys(name);
         addTagButton.click();
-        wait.until(ExpectedConditions.visibilityOf(successToaster));
+        elementToLoad(devicePage.successToaster);
         wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("tags-list-tag-" + name + "-item-delete")));
         driver.findElement(By.id("tags-list-tag-" + name + "-item-delete")).click();
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/TagsPage.java]: could we use a helper method? and declare an element so we can just click on it

File: src/main/java/pageObjects/TagsPage.java [hunk 6]
```diff
@@ -46,160 +44,121 @@ public TagsPage(WebDriver driver) {
...
-        wait.until(ExpectedConditions.visibilityOf(addTagField));
+        permissions.tagsButtonSidebarMenu.click();
+        elementToLoad(addTagField);
         addTagField.sendKeys(Constants.EDITABLE_TAG_NAME);
         addTagButton.click();
-        wait.until(ExpectedConditions.visibilityOf(successToaster));
+        wait.until(ExpectedConditions.visibilityOf(devicePage.successToaster));
         editButton.click();
         editField.sendKeys(Keys.CONTROL + "a");
         editField.sendKeys(Keys.DELETE);
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/TagsPage.java]: this can be made a helper method

File: src/main/java/pageObjects/TagsPage.java [hunk 7]
```diff
@@ -46,160 +44,121 @@ public TagsPage(WebDriver driver) {
...
     public void successfullyEditTag(String nameAfterEdit) {
         signInAsAdmin();
-        tagsButtonPage.click();
-        wait.until(ExpectedConditions.visibilityOf(addTagField));
+        permissions.tagsButtonSidebarMenu.click();
+        elementToLoad(addTagField);
         addTagField.sendKeys(Constants.EDITABLE_TAG_NAME);
         addTagButton.click();
-        wait.until(ExpectedConditions.visibilityOf(successToaster));
+        wait.until(ExpectedConditions.visibilityOf(devicePage.successToaster));
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/TagsPage.java]: use the helper method

**Review by BiancaJoca — APPROVED** [review]

File: src/main/java/pageObjects/TagsPage.java [hunk 8]
```diff
@@ -46,160 +44,121 @@ public TagsPage(WebDriver driver) {
...
+        devicePage.devicesButton.click();
+        elementToLoad(devicePage.addDeviceButton);
+        devicePage.addDeviceButton.click();
+        devicePage.nameField.sendKeys(randomName());
+        devicePage.serialNoField.sendKeys(randomNumber() + "");
+        devicePage.inventoryNoField.sendKeys(randomNumber() + "");
+        devicePage.invoiceNoField.sendKeys(randomNumber() + "");
+        devicePage.tagsList.sendKeys(Constants.EDITABLE_TAG_NAME);
+        devicePage.tagsListElement.click();
+        devicePage.saveDeviceButton.click();
```
Comment (CaprarClaudiu) [inline src/main/java/pageObjects/TagsPage.java]: you could make a helper method for this, or reuse the create device method

**Review by CaprarClaudiu — CHANGES_REQUESTED** [review]
Please do not refactor classes that are not necessary for this task (filter , permissions). Also you can do some refactoring to the TagsPageTests class also. (use helper methods, declare webelements to be used in the tests, the getText and required field code can be made helper methods)

**Review by ScorusMihai — APPROVED** [review]

**Review by CaprarClaudiu — APPROVED** [review]

**Pull request merged**